### ISSUE TITLE
"too many SQL variables" Error with pandas 0.23 - enable multivalues insert #19664 issue

### ISSUE THREAD

**Issue by tripkane** [pandas-dev/pandas#21103]
#### Code Sample, a copy-pastable example if possible

```python
#let's import packages to use
import numpy as np
import sqlite3 as sl
import pandas as pd
import datetime as dt
from sqlalchemy import create_engine
from sqlite3 import OperationalError



# #use pandas to import data
df = pd.DataFrame(np.arange(0,20000,1))
#create the engine to connect pandas with sqlite3
engine = create_engine('sqlite://')
#create connection
conn = engine.connect()
#convert df to sql table
df.to_sql('test',engine, if_exists='fail')
#print results
result = conn.execute("select * from test")
for row in result:
    print(row['index'])
conn.close()

#################
#my actual case#####
#################

#let's import packages to use
import numpy as np
import sqlite3 as sl
import pandas as pd
import datetime as dt
from sqlalchemy import create_engine
from sqlite3 import OperationalError

# #use pandas to import data
df = pd.read_excel(pth + 'CGM.xlsx', sheet_name=0)
#create the engine to connect pandas with sqlite3
engine = create_engine('sqlite:///H:\\Tests\\CO_db.db')
#create a connection to the database to run SQL Queries
conn = engine.connect()

#deletes empty space
conn.execute('VACUUM')

#create the CGM table unless it already exists
try:
    
    df.to_sql('CGM',engine, if_exists='fail', chunksize=100)
    
except ValueError as e:
    print(e)
    if str(e) == "Table 'CGM' already exists.":
        pass
    else:
        raise


```
#### Problem description

In pandas 0.22 I could write a dataframe to sql of reasonable size without error. Now I receive this error "OperationalError: (sqlite3.OperationalError) too many SQL variables". I am converting a dataframe with ~20k+ rows to sql. After looking around I suspect the problem lies in the limit set by sqlite3: SQLITE_MAX_VARIABLE_NUMBER which is set to 999 by default. This can apparently be changed by recompiling sqlite and adjusting this variable accordingly. I also noticed that adjusting the chunksize in DataFrame.to_sql has no effect perhaps confirming this is the root cause. I can confirm that for a df of length (rows) 499 this works. I can also confirm that this test version works with a row lenght of 20k and a chunksize of 499 inputted with df_to_sql works. My actual case is connecting to a local sqllite database and this is failing regardless of the chunksize adjusted to be 1 or anything else or without. #19664 

#### Expected  #Output
 
runfile('H:/Tests/Pandas_0.23_test.py', wdir='H:/Tests')
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.23.0
pytest: 3.5.1
pip: 10.0.1
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.14.3
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.4.0
sphinx: 1.7.4
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.3
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.3
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.4
lxml: 4.2.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

**Comment by kristang** [comment]
This is also an issue when connecting to MSSQL. It fails silently and will not stop the function or return an exception, but will throw one if you set "chunksize" manually.

**Comment by tripkane** [comment]
@kristang: I just played around a bit more. Have you tried a chunksize limit of 75 or less? My case is working for the local db based on this chunksize and strangely 499 for a sqlite db in memory, neither of which are the limits stated by sqlite. Have you tried different chunksizes in your case?

**Comment by kristang** [comment]
I tried 999, 1000 and 2100, but then just ended up downgrading to 0.22 to get on with my day.

**Comment by TomAugspurger** [comment]
@tripkane could I bother you to narrow down the original example a bit?  It should be minimal, and something that's appropriate for sticking in a unit test.

Could you also include the traceback you get?

**Comment by tripkane** [comment]
@kristang: :) fair enough maybe try these lower numbers this out when you have time

**Comment by tripkane** [comment]
@TomAugspurger: Hi Tom. Yep can do. So reduce it to my actual case (i.e. locally saved database i.e. not in memory and create the data from np.arange like the :memory: version?)

**Comment by TomAugspurger** [comment]
In-memory is fine if it reproduces the error.
http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports may be
helpful.

**Comment by tripkane** [comment]
@TomAugspurger: I have shortened it and updated the comments. Basically chunksize must be used and one needs to estimate this based on the data size of one row based on each individual case. This person implemented this solution with peewee: [https://stackoverflow.com/questions/35616602/peewee-operationalerror-too-many-sql-variables-on-upsert-of-only-150-rows-8-c](url)

**Comment by tripkane** [comment]
@jorisvandenbossche: You're welcome, happy to help.

**Comment by schettino72** [comment]
Since you guys are looking into this... Can you guys check if the change that caused this regression actually improved performance?

As I reported on #8953  this change actually made the operation slower for me. And IMO should be reverted...

**Comment by TomAugspurger** [comment]
Best to open a separate issue to track that. Could you do some profiling as
well, to see where the slowdown is?

**Comment by tripkane** [comment]
@TomAugspurger: I tested this quickly with essentially the same example as above for two different length df's 20k and 200k and compared chunksize = 1 with chunksize=999//(cols+1)  with cols = 10 and in both cases it was twice as fast when connected to a local sqlite db.

**Comment by schettino72** [comment]
@TomAugspurger I created issue #21146 to track this performance regression.
I did not profile it but did some experiments... my conclusion is that performance on 0.23 is better for a few columns but degrades quickly as the number of columns grow.

**Comment by cpcloud** [comment]
@pandas-dev/pandas-core How do people feel about reverting https://github.com/pandas-dev/pandas/pull/19664? It both breaks existing code and degrades performance significantly. Happy to put up the patch.

**Comment by TomAugspurger** [comment]
Reverting seems reasonable for now.

Is there anything we can salvage form #19664 so that users can opt into
multi-row inserts if they want?

**Comment by TomAugspurger** [comment]
cc @danfrankj

**Comment by cpcloud** [comment]
IMO pandas should allow more flexibility here: users pass in a function that will get called with the SQLAlchemy table and they can construct the DML themselves. The default DML construction function would be the previous working code. Then any such optimization that comes up in the future doesn't need yet another keyword argument. I wonder if there isn't already a way to do this without monkey patching.

**Comment by schettino72** [comment]
@cpcloud As of today there is no way to achieve this without monkey patching. 
I use a monkey patched version to use `COPY ... FROM`.

I could work on patch to make this more flexible. I suggest adding a parameter `insert_method` to `to_sql()`. Apart from the default (previous) implementation, we could add `multi-value` (the 0.23 version) and `COPY FROM`.

**Comment by jorisvandenbossche** [comment]
The way without monkeypatching is subclassing the table class and overriding the appropriate method, but that is not necessarily cleaner than monkeypatching as it private methods you are overriding anyway.

I like the proposed idea of allowing more flexibility. But we should think a bit about the exact interface we want (what does a user need to create the appropriate construct? The sqlalchemy table, the data (in which form?), the connection, ...).

**Comment by schettino72** [comment]
@jorisvandenbossche 

> The way without monkeypatching is subclassing the table class and overriding the appropriate method, but that is not necessarily cleaner than monkeypatching as it private methods you are overriding anyway.

`to_sql()` has no direct control of which table class will be used. So currently, subclassing is not enough, you MUST monkeypatch.

Also you would need to be careful with subclassing the table class as there are 2 implementations: SQLAlchemy and SQLite. 

Another problem of monkeypatching/subclassing (opposed to parameter) is that the same method has to be used globally by an application, and can not be fine tuned for each call of `to_sql()`.


> But we should think a bit about the exact interface we want (what does a user need to create the appropriate construct? The sqlalchemy table, the data (in which form?), the connection, ...).

I think just follow the API already used for `_execute_insert(self, conn, keys, data_iter)`

- a reference to table object
- a connection
- list of keys/columns
- and data_iter with values

That was enough for me to implement postgresql `COPY` without having to repeat Pandas specific processing of columns and conversion of values.

**Comment by jorisvandenbossche** [comment]
> to_sql() has no direct control of which table class will be used. So currently, subclassing is not enough, you MUST monkeypatch.

Yeah, for `to_sql` that is true, but in principle you *can* subclass and then use the methods of that class. With emphasis on "can", as I also don't really recommend it as I said in my previous comment.

Anyway, let's leave that discussion as I think we agree that we need some way to customize it (at least to switch between the single / multi values to solve the regression).

> I think just follow the API already used for _execute_insert(self, conn, keys, data_iter)

Would there be occasions where a user would want the actual dataframe instead of `data_iter`? 

@cpcloud what do you think of this interface?

@schettino72 One option (for the PR) would also be to implement this parameter, but leave the postgres-copy one as an example of how to use this (to not add an option that only works for a single engine for now; although given its superior speed, we should really think how to deal with this).

**Comment by danfrankj** [comment]
Hi all! My apologies for the regression introduced.  I like the solution in https://github.com/pandas-dev/pandas/pull/21199 much better. 

For context, the motivation for needing a multivalues insert was to interact with a presto database. In that case, a normal insert occurs row by row incurring a network call thus making it prohibitively slow. This may be the case for other analytic databases as well (redshift, bigquery).

But definitely we should not have this be the default so thank you @schettino72 for cleaning this up and again apologies.

**Comment by danfrankj** [comment]
Just verified that multivalues has this behavior for Redshift as well. 

```
In [60]: df = pd.DataFrame({'foo': np.arange(100)})

In [61]: conn.dialect
Out[61]: <sqlalchemy.dialects.postgresql.psycopg2.PGDialect_psycopg2 at 0x11d8e9748>

In [62]: conn.dialect.supports_multivalues_insert = False

In [63]: %time df.to_sql('test_no_multivalues_insert', con=conn, schema='default', index=False)
CPU times: user 21.4 ms, sys: 11.6 ms, total: 32.9 ms
Wall time: 38.7 s

In [64]: conn.dialect.supports_multivalues_insert = True

In [65]: %time df.to_sql('test_multivalues_insert', con=conn, schema='default', index=False)
CPU times: user 7.88 ms, sys: 2.38 ms, total: 10.3 ms
Wall time: 4.26 s
```

**Comment by jorisvandenbossche** [comment]
@schettino72 do you have time to work on the PR? 
Other short-term option for 0.23.1 is to revert the original change, and target the PR for 0.24.0

**Comment by schettino72** [comment]
> @schettino72 do you have time to work on the PR?

Sure. I was waiting for more feedback on API. From my side I guess just need to update docstrings and docs...

Should I target 0.24 or 0.23.1?

**Comment by TomAugspurger** [comment]
I think 0.23.1

**Comment by jorisvandenbossche** [comment]
Yes, we certainly want to do something for 0.23.1 (at a minimum reverse). But it would be nice to at once get the option in there as well. 
I am only not fully sure about the function interface (in case this needs more discussion for the exact API of the passed function). Maybe we could limit it for 0.23.1 only to adding `method='default'/'multi'`, and leave the ability to pass a function for 0.24.0

**Comment by gnfrazier** [comment]
Had the initial issue using 23.0 updating to 23.4 corrected the problem. Thanks for the fix.

**Comment by theSuiGenerisAakash** [comment]
On 1.0.1 and still facing the same issue

**Comment by vaibhav-chopda** [comment]
@theSuiGenerisAakash try removing "method='multi'" . This might help

**Comment by minderchen** [comment]
reduce to see whether it will work
chunksize=100

**Comment by theSuiGenerisAakash** [comment]
> @theSuiGenerisAakash try removing "method='multi'" . This might help

This defeats the purpose

**Comment by eromoe** [comment]
Happen on pandas 1.0.1 , this issue should reopen!

**Comment by AntoineWeber** [comment]
Getting the error on 1.1.5. For a wide dataframe (72 cols), I cannot go above a chunksize of 10.
